### ISSUE TITLE
Non-special Character Inserted Twice When Accepting Completion Option

### ISSUE THREAD

**Issue by rjkaes** [ms-jpq/coq_nvim#510]
Beginning with commit 5e95937974df23788ab99cb8bba6a10d2302457f whenever something other than `enter` is used to accept the selection, the non-`enter` character is inserted twice.

As an example, before 5e95937974df23788ab99cb8bba6a10d2302457f selecting the completion in the menu and typing `.` would accept the completion and insert a single `.` character. This is the behaviour for `ins-completion`. We are in state 1:

> 1. A complete match has been inserted, e.g., after using CTRL-N or CTRL-P.

Within state 1, the following is true:

> Any non-special character:
>                  Stop completion without changing the match and insert the
>                  typed character.

After 5e95937974df23788ab99cb8bba6a10d2302457f doing the same thing inserts *two* `.` characters after the accepted completion item from the menu.

**Comment by 9999years** [comment]
I'm seeing this bug, too.

**Comment by 9999years** [comment]
If I patch 5e95937974df23788ab99cb8bba6a10d2302457f to change `max(0, lo - 1)` to `max(0, lo + 1)`, then the case where inserting a character to accept the completion works correctly (the character is inserted and the cursor is after it). However, the case where Enter is used to accept the completion breaks: instead of moving to the end of the completion, the cursor doesn't move at all.

```diff
diff --git a/coq/server/edit.py b/coq/server/edit.py
index b6fb6418..dbc5af3e 100644
--- a/coq/server/edit.py
+++ b/coq/server/edit.py
@@ -436,12 +436,14 @@ def _restore(
         cur_row, cur_col = win_get_cursor(nvim, win=win)

         (_, lo), (_, hi) = m1.end, m2.begin
-        lo = max(0, lo - 1)
+        lo = max(0, lo + 1)

         binserted = encode(after)[lo:hi]
         inserted = decode(binserted)
```

**Comment by ms-jpq** [comment]
oh shit yeah my bad i messed around with the off by 1 sruff ill fix after work

**Comment by ms-jpq** [comment]
ed9d30c64e84a0eca1f0749c7eb169f204724404 fix

**Comment by 9999years** [comment]
Looks like that fixes the case when inserting a character, but hitting enter still leaves my cursor where it was (rather than at the end of the completion).

```
# | is cursor
Line:     hello-|
Select completion "hello-world" with Enter
Line:     hello-|world
Expected: hello-world|
```

**Comment by ms-jpq** [comment]
oh i need coffee lol

**Comment by ms-jpq** [comment]
ok i fixed it

**Comment by 9999years** [comment]
Can confirm this is fixed in (I think?) 6cbd89e1270aa84d88843b702aa33e37529fa100.

**Comment by rjkaes** [comment]
Appears to be working for me as expected!  Thanks for the fix!

Great project! 😄
